### PR TITLE
feat(lightspeed): create lightspeed backend

### PR DESCRIPTION
**Pull request by yangcao77**
create a basic implementation of lightspeed backend plugin
introduced `v1/query` endpoint as in designed doc: https://docs.google.com/document/d/1nzW-crxBZoayWDWlZbtk0q6bmWFPy9qJ8Yun_zNOvRk/edit#heading=h.25xpl866izc8

tested with the current lightspeed frontend plugin, update the `LightSpeedPage` to:
```
      const result = await lightspeedApi.createChatCompletions(prompt);
      setChunkIndex(index => index + 1);
      setCompletions(c => {
            c[prompt] = result?.response;
            return c;
      });
```

and update the `LightspeedProxyClient` to be following:
```
import {
  ConfigApi,
  createApiRef,
  IdentityApi,
  DiscoveryApi,
  FetchApi
} from '@backstage/core-plugin-api';



export type LightspeedAPI = {
  createChatCompletions: (
    message: string,
  ) => Promise<any | { error: any }>; 
};

export const lightspeedApiRef = createApiRef<LightspeedAPI>({
  id: 'plugin.lightspeed.service',
});

export type Options = {
  configApi: ConfigApi;
  identityApi: IdentityApi;
  discoveryApi: DiscoveryApi;
  fetchApi: FetchApi;
};

export class LightspeedProxyClient implements LightspeedAPI {
  // @ts-ignore
  private readonly configApi: ConfigApi;
  private readonly identityApi: IdentityApi;
  private readonly discoveryApi: DiscoveryApi;
  private readonly fetchApi: FetchApi;

  constructor(options: Options) {
    this.configApi = options.configApi;
    this.identityApi = options.identityApi;
    this.discoveryApi = options.discoveryApi;
    this.fetchApi = options.fetchApi;

  }

  async createChatCompletions(message: string) {
    try {
      const baseUrl = await this.discoveryApi.getBaseUrl('lightspeed');
      const data = {
        model: 'ibm-granite-8b-code-instruct', // model name
        conversation_id: "user1+1q2w3e4r-abcd1234", // dummy conversation_id
        serverURL: `${this.configApi.getString('backend.baseUrl')}/api/proxy/lightspeed/api`,
        query: message,
      };
      const resp = await this.fetchApi.fetch(`${baseUrl}/v1/query`, {
        method: 'POST',
        body: JSON.stringify(data),
        headers: {
          'Content-Type': 'application/json',
        },
      });
      const respData = await resp.json();

      return respData;
    } catch (error: any) {
      return { error: error.message };
    }

  }
}

```

Screenshot 2024-08-30 at 5 07 05 PM

### CODE REVIEW

**Comment by yangcao77** [comment]
@rohitkrai03  PTAL. This is the first pass of the lightspeed backend, which only implements the chat completion. https://docs.google.com/document/d/1nzW-crxBZoayWDWlZbtk0q6bmWFPy9qJ8Yun_zNOvRk/edit#heading=h.25xpl866izc8

File: plugins/lightspeed-backend/src/service/router.ts
Comment (dzemanov) [inline plugins/lightspeed-backend/src/service/router.ts:82]:
```suggestion
        const errormsg = `Error fetching completions from ${serverURL}: ${error}`
        logger.error(errormsg);
        response.status(500).json({ error: errormsg});
      }
```

If openapi won't be needed, then this can be simplified.

File: plugins/lightspeed-backend/src/service/router.ts
Comment (dzemanov) [inline plugins/lightspeed-backend/src/service/router.ts:13]:
```suggestion
import { HumanMessage } from "@langchain/core/messages";
import { ChatPromptTemplate, MessagesPlaceholder } from "@langchain/core/prompts";
```

You can use pre-commit hooks to lint your code, here is more info: https://github.com/janus-idp/backstage-plugins/blob/main/CONTRIBUTING.md#lint

File: plugins/lightspeed-backend/README.md
Comment (dzemanov) [inline plugins/lightspeed-backend/README.md:78]: Would update to new configuration if you will have time. This is not needed anymore.

File: plugins/lightspeed-backend/src/service/router.test.ts
Comment (dzemanov) [inline plugins/lightspeed-backend/src/service/router.test.ts:72]: Maybe a different description here as streaming is set to false for now

File: plugins/lightspeed-backend/src/service/router.test.ts
Comment (dzemanov) [inline plugins/lightspeed-backend/src/service/router.test.ts:86]:
```suggestion
    it('returns 400 if conversation_id is missing', async () => {
```

File: plugins/lightspeed-backend/package.json
Comment (yangcao77) [reply plugins/lightspeed-backend/package.json]: updated

File: plugins/lightspeed-backend/src/service/router.ts
Comment (yangcao77) [reply plugins/lightspeed-backend/src/service/router.ts]: updated

File: plugins/lightspeed-backend/README.md
Comment (yangcao77) [reply plugins/lightspeed-backend/README.md]: updated readme

File: plugins/lightspeed-backend/README.md
Comment (yangcao77) [reply plugins/lightspeed-backend/README.md:98]: yes. The proxy is for frontend plugins. for the lightspeed backend, we expect frontend to send the model & baseURL to backend. 
However, I listed this configuration in readme, since backend will need the authorization header to be set to successfully call the llm server. 

Later on, I'm thinking we can have a single readme file for both lightspeed frontend and backend to document the proper setup. for example the [orchestrator readme](https://github.com/janus-idp/backstage-plugins/blob/main/plugins/orchestrator-backend/README.md)

Comment (dzemanov) [reply plugins/lightspeed-backend/README.md:98]: Understood :). Maybe you can add information that this is composed of backend and frontend lightspeed plugin and these are instructions for the frontend part, similiar as ocm [README](https://github.com/janus-idp/backstage-plugins/tree/main/plugins/ocm) has instructions for ocm-backend as well. But that is up to you.

**Comment by yangcao77** [comment]
> @rohitkrai03 PTAL. This is the first pass of the lightspeed backend, which only implements the chat completion. https://docs.google.com/document/d/1nzW-crxBZoayWDWlZbtk0q6bmWFPy9qJ8Yun_zNOvRk/edit#heading=h.25xpl866izc8

@rohitkrai03  PTAL when you have time. so we can unblocked from implementing the chat history persistence. Thanks!

**Comment by rohitkrai03** [comment]
@yangcao77 Apart from the changes we discussed around backend config and APIs, this PR is good to go from my side.

**Review by rohitkrai03 — APPROVED** [review]
/lgtm

**Comment by yangcao77** [comment]
@rohitkrai03 can you help add back the lgtm label again? thanks!
I pushed a new commit to fix the prettier check. but the CI build cannot be triggered unless it's approved

**Review by nickboldt — CHANGES_REQUESTED** [review]
tests are failing, please review 

Also I'm setting this to blocked until we decide if it's OK to put this into a deprecated repo and end up with your plugin published under a defunct namespace `@janus-idp` instead of something like `@ansible-lightspeed` or `@redhat-developer`

File: plugins/lightspeed-backend/package.json
Comment (nickboldt) [inline plugins/lightspeed-backend/package.json:7]: once merged, you'll have to switch this to false in order to do releases to npmjs.com under the janus-idp namespace. See https://issues.redhat.com/browse/RHIDP-4029

**Review by nickboldt — APPROVED** [review]
Both @christophe-f  and @durandom have agreed that we should include this plugin in janus until we have fleshed out the requirements for a new rhdh-plugins repo. 

Therefore I've removed my previous objection and will deal with the MSR build/publish failures that will inevitably happen when we merge this - see https://issues.redhat.com/browse/RHIDP-4029

Please fix your failing tests to proceed:

https://github.com/janus-idp/backstage-plugins/actions/runs/10853505056/job/30182556594?pr=2115

```
src/plugin.ts:5:1 - error TS6133: 'catalogServiceRef' is declared but its value is never read.

5 import { catalogServiceRef } from '@backstage/plugin-catalog-node/alpha';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/service/router.test.ts:2:1 - error TS6192: All imports in import declaration are unused.

2 import { mockCredentials, mockServices } from '@backstage/backend-test-utils';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/service/standaloneServer.ts:6:1 - error TS61[92](https://github.com/janus-idp/backstage-plugins/actions/runs/10853505056/job/30182556594?pr=2115#step:4:104): All imports in import declaration are unused.

6 import { mockCredentials, mockServices } from '@backstage/backend-test-utils';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 3 errors in 3 files.

Errors  Files
     1  src/plugin.ts:5
     1  src/service/router.test.ts:2
     1  src/service/standaloneServer.ts:6
```

**Comment by yangcao77** [comment]
@rohitkrai03 @nickboldt 
the tests finally passed. Do you mind helping add back the lgtm label again please? Thanks!

File: plugins/lightspeed-backend/package.json
Comment (yangcao77) [reply plugins/lightspeed-backend/package.json:7]: according to the discussion, see comments in https://issues.redhat.com/browse/RHIDP-4029
we decided to keep this plugin as private with `0.x` version, and will rely on manual releasing upon request.

**Pull request merged**
